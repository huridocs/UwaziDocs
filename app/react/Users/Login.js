import PropTypes from 'prop-types';
import React from 'react';
import { browserHistory } from 'react-router';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Field, Form, actions as formActions } from 'react-redux-form';

import { t } from 'app/I18N';
import { reconnectSocket } from 'app/socket';
import RouteHandler from 'app/App/RouteHandler';
import ShowIf from 'app/App/ShowIf';
import { reloadThesauris } from 'app/Thesauris/actions/thesaurisActions';

import auth from 'app/Auth';

const reloadHome = () => {
  window.location.reload();
};

export class Login extends RouteHandler {
  constructor(props, context) {
    super(props, context);
    this.state = { error: false, recoverPassword: false };
    this.submit = this.submit.bind(this);
  }

  submit(credentials) {
    if (this.state.recoverPassword) {
      return this.recoverPassword(credentials.username);
    }

    return this.login(credentials);
  }

  static requestState() {
    return Promise.resolve({});
  }

  recoverPassword(email) {
    this.setState({ recoverPassword: false, error: false });
    this.props.reset('login.form');
    return this.props.recoverPassword(email);
  }

  login(credentials) {
    return this.props.login(credentials)
    .then(() => {
      if (this.props.private) {
        browserHistory.push('/');
        reloadHome();
        return;
      }
      reconnectSocket();
      this.props.reloadThesauris();
      browserHistory.push('/');
    })
    .catch(() => {
      this.setState({ error: true });
    });
  }

  setRecoverPassword() {
    this.props.reset('login.form');
    this.setState({ recoverPassword: true, error: false });
  }

  setLogin() {
    this.props.reset('login.form');
    this.setState({ recoverPassword: false, error: false });
  }

  render() {
    return (
      <div className="content login-content">
        <div className="row">
          <div className="col-xs-12 col-sm-4 col-sm-offset-4">
            <h1 className="login-title">
              <img src="/public/logo.svg" title="uwazi" alt="uwazi"/>
            </h1>
            <Form onSubmit={this.submit} model="login.form">
              <div className={`form-group login-email${this.state.error ? ' has-error' : ''}`}>
                <Field model="login.form.username">
                  <label className="form-group-label" htmlFor="username">
                    {this.state.recoverPassword ? t('System', 'Email') : t('System', 'User')}
                  </label>
                  <input type="text" name="username" id="username" className="form-control"/>
                </Field>
              </div>
              <div className={`form-group login-password ${this.state.error ? 'has-error' : ''}${this.state.recoverPassword ? ' is-hidden' : ''}`}>
                <label className="form-group-label" htmlFor="password">{t('System', 'Password')}</label>
                <Field model="login.form.password">
                  <input type="password" name="password" id="password" className="form-control"/>
                </Field>
                <div className="form-text">
                  <span className="required">{t('System', 'Login failed')} - </span>
                  <a
                    title={t('System', 'Forgot Password?', null, false)}
                    onClick={this.setRecoverPassword.bind(this)}
                    className={(this.state.error ? 'label-danger' : '')}
                  >
                    {t('System', 'Forgot Password?')}
                  </a>
                </div>
              </div>
              <p>
                <button type="submit" className={`btn btn-block btn-lg ${this.state.recoverPassword ? 'btn-success' : 'btn-primary'}`}>
                  {this.state.recoverPassword ? t('System', 'Send recovery email') : t('System', 'Login button', 'Login')}
                </button>
              </p>
              <ShowIf if={this.state.recoverPassword}>
                <div className="form-text">
                  <a
                    title={t('System', 'Cancel', null, false)}
                    onClick={this.setLogin.bind(this)}
                  >
                    {t('System', 'Cancel')}
                  </a>
                </div>
              </ShowIf>
            </Form>
          </div>
        </div>
      </div>
    );
  }
}

Login.propTypes = {
  login: PropTypes.func,
  recoverPassword: PropTypes.func,
  reloadThesauris: PropTypes.func
};

export function mapStateToProps({ settings }) {
  return {
    private: settings.collection.get('private')
  };
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators({
    login: auth.actions.login,
    recoverPassword: auth.actions.recoverPassword,
    reset: formActions.reset,
    reloadThesauris
  }, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(Login);
