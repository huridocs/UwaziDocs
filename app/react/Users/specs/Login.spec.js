import React from 'react';
import {shallow} from 'enzyme';
import {browserHistory} from 'react-router';

import {Login} from '../Login.js';
import Alert from 'app/components/Elements/Alert.js';

describe('Login', () => {
  let component;
  let instance;
  let props;

  let render = () => {
    component = shallow(<Login {...props}/>);
    instance = component.instance();
  };

  beforeEach(() => {
    props = {login: jasmine.createSpy('login').and.returnValue(Promise.resolve())};
    render();
  });

  describe('on instance', () => {
    it('should set state with blank username and password', () => {
      expect(instance.state.credentials).toEqual({username: '', password: ''});
    });

    it('should set state with error false', () => {
      expect(instance.state.error).toEqual(false);
    });
  });

  describe('render', () => {
    describe('when there is NOT an error', () => {
      it('should NOT display an error message', () => {
        expect(component.find(Alert).length).toBe(0);
      });
    });

    describe('when there is an error', () => {
      it('should display an error message', () => {
        component.setState({error: true});
        expect(component.find(Alert).length).toBe(1);
      });
    });
  });

  describe('submit()', () => {
    describe('on response success', () => {
      it('should set error false', (done) => {
        instance.state.error = true;

        instance.submit(new Event('submit'))
        .then(() => {
          expect(instance.state.error).toBe(false);
          done();
        })
        .catch(done.fail);
      });

      it('should go to home', (done) => {
        spyOn(browserHistory, 'push');
        instance.submit(new Event('submit'))
        .then(() => {
          expect(browserHistory.push).toHaveBeenCalledWith('/');
          done();
        })
        .catch(done.fail);
      });
    });

    describe('on response failure', () => {
      it('should set error true', (done) => {
        props.login = jasmine.createSpy('login').and.returnValue(Promise.reject());
        render();

        instance.submit(new Event('submit'))
        .then(() => {
          expect(instance.state.error).toBe(true);
          done();
        })
        .catch(done.fail);
      });
    });
  });
});
