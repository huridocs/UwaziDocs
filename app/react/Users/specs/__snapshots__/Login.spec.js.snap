// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login on instance should render the component with the login form 1`] = `
<div
  className="content login-content"
>
  <div
    className="row"
  >
    <div
      className="col-xs-12 col-sm-4 col-sm-offset-4"
    >
      <h1
        className="login-title"
      >
        <img
          alt="uwazi"
          src="/public/logo.svg"
          title="uwazi"
        />
      </h1>
      <LocalForm
        getDispatch={[Function]}
        initialState={Object {}}
        model="loginForm"
        onSubmit={[Function]}
      >
        <div
          className="form-group login-email"
        >
          <Modeled(Connect(Field))
            model=".username"
          >
            <label
              className="form-group-label"
              htmlFor="username"
            >
              <Translate>
                User
              </Translate>
            </label>
            <input
              className="form-control"
              id="username"
              name="username"
              type="text"
            />
          </Modeled(Connect(Field))>
        </div>
        <div
          className="form-group login-password "
        >
          <label
            className="form-group-label"
            htmlFor="password"
          >
            <Translate>
              Password
            </Translate>
          </label>
          <Modeled(Connect(Field))
            model=".password"
          >
            <input
              className="form-control"
              id="password"
              name="password"
              type="password"
            />
          </Modeled(Connect(Field))>
          <div
            className="form-text"
          >
            <span
              className="button forgot-password "
              onClick={[Function]}
              title="Forgot Password?"
            >
              <Translate>
                Forgot Password?
              </Translate>
            </span>
          </div>
        </div>
        <p>
          <button
            className="btn btn-block btn-lg btn-primary"
            type="submit"
          >
            <Translate>
              Login
            </Translate>
          </button>
        </p>
      </LocalForm>
    </div>
  </div>
</div>
`;

exports[`Login submit() on response failure when authorization conflict (2fa required) should not set error and flag "tokenRequired" 1`] = `
<div
  className="content login-content"
>
  <div
    className="row"
  >
    <div
      className="col-xs-12 col-sm-4 col-sm-offset-4"
    >
      <h1
        className="login-title"
      >
        <img
          alt="uwazi"
          src="/public/logo.svg"
          title="uwazi"
        />
      </h1>
      <LocalForm
        getDispatch={[Function]}
        initialState={Object {}}
        model="loginForm"
        onSubmit={[Function]}
      >
        <div
          className="form-group login-token"
        >
          <h5>
            <Translate
              context="System"
            >
              Two-step verification
            </Translate>
          </h5>
          <Modeled(Connect(Field))
            model=".token"
          >
            <label
              className="form-group-label"
              htmlFor="token"
            >
              <Translate
                context="System"
              >
                Authentication code
              </Translate>
            </label>
            <input
              className="form-control"
              id="token"
              name="token"
              type="text"
            />
            <div
              className="form-text"
            >
              <p />
              <p>
                <Translate>
                  Open the two-factor Authenticator app on your device
                </Translate>
                 
                <br />
                <Translate>
                  to view your authentication code and verify your identity.
                </Translate>
              </p>
              <p>
                <span
                  className="button forgot-password "
                  onClick={[Function]}
                >
                  <Translate>
                    Return to login
                  </Translate>
                </span>
              </p>
            </div>
          </Modeled(Connect(Field))>
        </div>
        <p>
          <button
            className="btn btn-block btn-lg btn-primary"
            type="submit"
          >
            <Translate>
              Verify
            </Translate>
          </button>
        </p>
      </LocalForm>
    </div>
  </div>
</div>
`;

exports[`Login submit() on response failure when authorization failure should set error2fa on token failure, and reset loginForm.token value 1`] = `
<div
  className="content login-content"
>
  <div
    className="row"
  >
    <div
      className="col-xs-12 col-sm-4 col-sm-offset-4"
    >
      <h1
        className="login-title"
      >
        <img
          alt="uwazi"
          src="/public/logo.svg"
          title="uwazi"
        />
      </h1>
      <LocalForm
        getDispatch={[Function]}
        initialState={Object {}}
        model="loginForm"
        onSubmit={[Function]}
      >
        <div
          className="form-group login-token has-error"
        >
          <h5>
            <Translate
              context="System"
            >
              Two-step verification
            </Translate>
          </h5>
          <Modeled(Connect(Field))
            model=".token"
          >
            <label
              className="form-group-label"
              htmlFor="token"
            >
              <Translate
                context="System"
              >
                Authentication code
              </Translate>
            </label>
            <input
              className="form-control"
              id="token"
              name="token"
              type="text"
            />
            <div
              className="form-text"
            >
              <p>
                <Connect(Icon)
                  icon="exclamation-triangle"
                />
                <span>
                  <Translate
                    context="System"
                  >
                    Two-factor verification failed
                  </Translate>
                </span>
              </p>
              <p>
                <Translate>
                  Open the two-factor Authenticator app on your device
                </Translate>
                 
                <br />
                <Translate>
                  to view your authentication code and verify your identity.
                </Translate>
              </p>
              <p>
                <span
                  className="button forgot-password label-danger"
                  onClick={[Function]}
                >
                  <Translate>
                    Return to login
                  </Translate>
                </span>
              </p>
            </div>
          </Modeled(Connect(Field))>
        </div>
        <p>
          <button
            className="btn btn-block btn-lg btn-primary"
            type="submit"
          >
            <Translate>
              Verify
            </Translate>
          </button>
        </p>
      </LocalForm>
    </div>
  </div>
</div>
`;
